### PR TITLE
spring-boot-security1

### PR DESCRIPTION
**Pull request by DojiDoe**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -94,6 +94,12 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.1.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.4</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/Role.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+@Setter
+@Table(name = "roles")
+@RequiredArgsConstructor
+public class Role implements GrantedAuthority {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName role = RoleName.ROLE_USER;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Role.java]:
```suggestion
    private RoleName role;
```

File: src/main/java/mate/academy/bookstore/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "roles")
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Role.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 4]
```diff
@@ -26,4 +34,44 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 5]
```diff
@@ -26,4 +34,44 @@ public class User {
...
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find user by email"));
```

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 7]
```diff
@@ -1,26 +1,34 @@
...
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         if (userRepository.existsByEmail(requestDto.getEmail())) {
             throw new RegistrationException("Can't register user");
         }
+        requestDto.setPassword(passwordEncoder.encode(requestDto.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java]: set password to User user, not dto

File: src/main/java/mate/academy/bookstore/service/UserServiceImpl.java [hunk 8]
```diff
@@ -1,26 +1,34 @@
...
 
 @Service
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserServiceImpl.java:25]:
```suggestion
    @Override
    @Transactional
```

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role
+                  type: ENUM('ROLE_ADMIN','ROLE_USER','ROLE_MANAGER')
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: bad approach to add enum here, it will cause some problems during modify it

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**